### ISSUE TITLE
Proposal: initialize fields at the declaration site

### ISSUE THREAD

**Issue by nate-thegrate** [flutter/flutter#153644]
#### before
```dart
class _MyState extends State<MyWidget> with SingleTickerProviderMixin {
  late final int _value;
  AnimationController? _controller;

  @override
  void initState() {
    _value = widget.value;
    _controller = AnimationController(
      duration: Durations.medium1,
      vsync: this,
    );
  }

  // ...
}
```

#### after
```dart
class _MyState extends State<MyWidget> with SingleTickerProviderMixin {
  late final int _value = widget.value;

  late final AnimationController _controller = AnimationController(
    duration: Durations.medium1,
    vsync: this,
  );

  // ...
}
```

<br>

This has a few advantages:
1. More concise
2. Less jumping around to see how fields are initialized
3. Less prone to late initialization errors
4. The [prefer final](https://dart.dev/diagnostics/prefer_final_fields) & [unnecessary nullable](https://dart.dev/diagnostics/unnecessary_nullable_for_final_variable_declarations) lints will kick in

**Comment by moffatman** [comment]
I thought it was this way because of some code path where the state object could be created. But never actually mounted. Therefore would not be dispose()ed. However I think you're right upon looking into it.... this should be safe

**Comment by jonahwilliams** [comment]
This is not always a safe change. late finals are not initialized until accessed. Simple example, the following code will never update.

```dart
class FooState extends State<Foo> {
  late final timer = Timer.periodic(Duration(milliseconds: 10), (timer) {
    setState(() {
      count++;
    });
  });
  int count = 0;

  Widget build(_) => Text('Count $cound');
}

```

**Comment by nate-thegrate** [comment]
Yeah, forgot to mention that, but it's a good point—there was one time when I tried subscribing to a Firebase stream using a `late final` variable, and never actually subscribed to it 😒

**Comment by nate-thegrate** [comment]
> some code path where the state object could be created but never actually mounted… However I think you're right upon looking into it

I wanted to double-check as well 🙂
Fortunately, it does look like every `State` object is mounted when it's created:

https://github.com/flutter/flutter/blob/1633c47dda49e5470989d0518e65fe113adbfcda/packages/flutter/lib/src/widgets/framework.dart#L4483-L4488

<br>

But recently I learned that I'd missed something important—there are actually 2 potential dangers involving `late` fields.

#### late initialization errors

Happens when a `late` variable is referenced before being initialized; can be avoided by initializing at the declaration site.

#### calling `createTicker()` inside `dispose()`

`AnimationController` calls `vsync.createTicker()` in its constructor, so if a `late` controller isn't referenced until it's being disposed of, it can result in an error.

For example, you could have a method that references the controller, but the controller won't be initialized until if/when the method is called:

```dart
class _MyState extends State<MyWidget> with TickerProviderMixin {
  late final _controller = AnimationController(vsync: this, duration: Durations.long1)
    ..addListener(() => setState(() {}));

  void _open() {
    _controller.forward();
  }

  @override
  void dispose() {
    _controller.dispose(); // dangerous
    super.dispose();
  }

  // ...
}
```
